File: app/src/main/java/hr/caellian/notestream/gui/ActivitySearch.java
package hr.caellian.notestream.gui;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.widget.DrawerLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Objects;

import hr.caellian.notestream.NoteStream;
import hr.caellian.notestream.R;
import hr.caellian.notestream.data.Playlist;
import hr.caellian.notestream.data.playable.Playable;
import hr.caellian.notestream.data.playable.PlayableYouTube;
import hr.caellian.notestream.data.youtube.YouTubeFetcher;
import hr.caellian.notestream.gui.fragments.FragmentItemPlayable;

/**
 * Created by caellyan on 25/06/17.
 */

public class ActivitySearch extends NavigationActivity {

    boolean showingResults = false;
    public final Playlist resultsPlaylist = Playlist.get("searchResults");

    private int fragmentCounter = 0;
    protected final ArrayList<FragmentItemPlayable> searchItems = new ArrayList<>();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(null);
        setContentView(R.layout.activity_search);

        navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        findViewById(R.id.buttonClearSearch).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ((EditText) findViewById(R.id.exitTextSearch)).setText("");
                resultsPlaylist.clear();
                refreshSearchResults();
            }
        });

        ((EditText) findViewById(R.id.exitTextSearch)).addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {}

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                final String searched = charSequence.toString();

                resultsPlaylist.clear();
                resultsPlaylist.add(NoteStream.getInstance().library.localMusic.filtered(searched));

                new AsyncTask<Void, Void, Void>() {
                    @Override
                    protected Void doInBackground(Void[] params) {
                        for (String youtubeID : YouTubeFetcher.searchFor(searched)) {
                            resultsPlaylist.add(new PlayableYouTube(youtubeID));
                            Toast.makeText(ActivitySearch.this, youtubeID, Toast.LENGTH_SHORT).show();
                        }
                        resultsPlaylist.add(new PlayableYouTube("8xg3vE8Ie_E"));
                        refreshSearchResults();
                        return null;
                    }
                }.execute();
            }

            @Override
            public void afterTextChanged(Editable editable) {}
        });
    }

    public void refreshSearchResults() {
        FragmentManager fm = getFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        for (Fragment searchItem : searchItems) {
            ft.remove(fm.findFragmentById(searchItem.getId()));
        }
        ft.commit();

        ft = fm.beginTransaction();
        searchItems.clear();
        for (Playable playable: resultsPlaylist) {
            FragmentItemPlayable fragment = FragmentItemPlayable.newInstance(playable, resultsPlaylist);
            searchItems.add(fragment);
            ft.add(R.id.foundContent, fragment , "playable-" + fragmentCounter++ );
        }
        ft.commit();
    }

    @Override
    public void onBackPressed() {
        if (!Objects.equals(((EditText) findViewById(R.id.exitTextSearch)).getText().toString(), "")) {
            ((EditText) findViewById(R.id.exitTextSearch)).setText("");
            resultsPlaylist.clear();
            refreshSearchResults();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected DrawerLayout getDrawerLayout() {
        return (DrawerLayout) findViewById(R.id.search_layout);
    }
}

File: app/src/main/java/hr/caellian/notestream/gui/ActivitySearch.kt
/*
 * Copyright (C) 2018 Tin Svagelj
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
 * persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package hr.caellian.notestream.gui

import android.os.Bundle
import android.os.Handler
import android.support.v4.widget.DrawerLayout
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import hr.caellian.notestream.NoteStream
import hr.caellian.notestream.R
import hr.caellian.notestream.data.playable.Playable
import hr.caellian.notestream.data.playable.PlayableYouTube
import hr.caellian.notestream.data.playlist.PlaylistIterator
import hr.caellian.notestream.data.playlist.PlaylistYouTube
import hr.caellian.notestream.data.youtube.YouTubeFetcher
import hr.caellian.notestream.gui.fragments.FragmentItemPlayable
import hr.caellian.notestream.gui.fragments.FragmentPlaylistTile
import hr.caellian.notestream.gui.fragments.FragmentTileYouTube
import java.util.*

class ActivitySearch : NavigationActivity() {
    internal var showingResults = false
    val youtubePlaylists = mutableListOf<PlaylistYouTube>()
    val youtubeResults = mutableListOf<PlayableYouTube>()
    val localResults = mutableListOf<Playable>()

    private var fragmentCounter = 0

    private val searchHandler = Handler()

    override val drawerLayout: DrawerLayout?
        get() = findViewById<View>(R.id.search_layout) as DrawerLayout


    private val youtubeSearch = Runnable {
        val query = findViewById<EditText>(R.id.textEditSearch)?.text?.toString()!!
        youtubePlaylists.clear()
        youtubeResults.clear()

        youtubePlaylists += YouTubeFetcher.searchPlaylists(query)
        youtubeResults += YouTubeFetcher.searchVideos(query)

        refreshSearchResults()
        findViewById<EditText>(R.id.youtubePlaylistArea)?.visibility = View.VISIBLE
        findViewById<EditText>(R.id.youtubePlayableArea)?.visibility = View.VISIBLE
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(null)
        setContentView(R.layout.activity_search)

        navigationView = findViewById(R.id.nav_view)
        navigationView?.setNavigationItemSelectedListener(this)

        findViewById<View>(R.id.buttonClearSearch)?.setOnClickListener {
            (findViewById<View>(R.id.textEditSearch) as? EditText)?.setText("")
            localResults.clear()
            findViewById<EditText>(R.id.youtubePlaylistArea)?.visibility = View.GONE
            findViewById<EditText>(R.id.youtubePlayableArea)?.visibility = View.GONE
            refreshSearchResults()
        }

        findViewById<EditText>(R.id.textEditSearch)?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                val searched = charSequence.toString()
                findViewById<EditText>(R.id.youtubePlaylistArea)?.visibility = View.GONE
                findViewById<EditText>(R.id.youtubePlayableArea)?.visibility = View.GONE

                localResults.clear()
                localResults += NoteStream.instance.data.localMusic.filtered(searched)
                searchHandler.removeCallbacks(youtubeSearch)
                searchHandler.postDelayed(youtubeSearch,2000)
            }

            override fun afterTextChanged(editable: Editable) {
                refreshSearchResults()
            }
        })
    }

    fun refreshSearchResults() {
        findViewById<LinearLayout>(R.id.localPlayable)?.removeAllViewsInLayout()
        findViewById<LinearLayout>(R.id.youtubePlayable)?.removeAllViewsInLayout()
        findViewById<LinearLayout>(R.id.localPlayable)?.removeAllViewsInLayout()

        val fm = fragmentManager
        val ft = fm.beginTransaction()

        for (playable in localResults) {
            val fragment = FragmentItemPlayable.newInstance(playable, PlaylistIterator(localResults))
            ft.add(R.id.localPlayable, fragment, "resultFragment-" + fragmentCounter++)
        }

        for (playable in youtubeResults) {
            val fragment = FragmentItemPlayable.newInstance(playable, PlaylistIterator(youtubeResults))
            ft.add(R.id.youtubePlayable, fragment, "resultFragment-" + fragmentCounter++)
        }

        for (playlist in youtubePlaylists) {
            val fragment = FragmentTileYouTube.create(playlist)
            ft.add(R.id.localPlayable, fragment, "resultFragment-" + fragmentCounter++)
        }
        ft.commit()
    }

    override fun onBackPressed() {
        if ((findViewById<View>(R.id.textEditSearch) as EditText?)?.text.toString() != "") {
            (findViewById<View>(R.id.textEditSearch) as EditText?)?.setText("")
            localResults.clear()
            refreshSearchResults()
        } else {
            super.onBackPressed()
        }
    }
}
